### DISCUSSION TITLE
Mongodb replica set with authentication enabled

### DISCUSSION THREAD

**Discussion by milindagashe** [typegoose/mongodb-memory-server#634] [Q&A]
## Versions

- NodeJS: 16
- mongodb-memory-server-*: 8.4.2
- mongodb(the binary version): 
- mongodb(the js package): 4.5.0
- mongoose: 6.2.10
- system:  Windows, Linux

package: mongo-memory-server 

## What is your question?
I am using below piece of code to create a replica set and it works fine with I am accessing the mongodb from the same server. From the different server, it says connection refused. I understand we have to add authentication to be able to gain access to mongodb from another server. Could you please suggest how to do it? I tried various things, but somehow all my attempts failed.

```ts
const replset = await MongoMemoryReplSet.create({
      instanceOpts: [{
        port: 57398
      }],
      replSet: {
        count: 1,
        dbName: 'TestDB'
      }
    })
    const uri = replset.getUri()
```

**Comment by hasezoey** [comment]
i think your problem is 1 or 2 of the following 2 options:
- you are not keeping the server alive long enough (the instance that MMS handles will only be alive when the nodejs process did not exit and `.stop` has not been called)
- if you want to access it from "not the local machine", you will have to [bind its ip](https://www.mongodb.com/docs/manual/core/security-mongodb-configuration/), which can be easily done in MMS

for replsets:
```ts
const replset = await MongoMemoryReplSet.create({
  replSet: {
    ip: "0.0.0.0" // change this ip to something more secure if wanted, see linked documentation for more
  }
});
```

for single instances:
```ts
const instance = await MongoMemoryServer.create({
  instance: {
    ip: "0.0.0.0" // change this ip to something more secure if wanted, see linked documentation for more
  }
});
```

PS: i will move this issue to a discussion

**Reply by milindagashe** [reply]
Hello, I checked both above things. Nodejs process which creates replica set does exist while other process tries to access the DB.
I also used '0.0.0.0' as suggested above, but in vain. 
To give some more info, I am facing this issue when I am using docker-compose. In docker compose, every service is given a unique IP address which can be accessed only within the network that docker-compose creates by default. In docker-compose config, there are 2 services, one which is responsible for creating a mongo memory server replicaset and the other which is using this replicaset. Both the processes are up and running. After inspecting the logs I found below:
service 1 (which creates db replicaset): IP address: 172.18.0.9
service 2(which consumes the db): IP address: 172.18.0.7
Service 1 is delayed by 20 seconds, so it means it only creates a db replicaset after 20 seconds after it starts. Service 2 starts to connect to the db right after it starts. So I am expecting "no connection" for the first 20 seconds and later Service 2 should establish the connection.
Service 2 logs look like below:
For the first 20 seconds, it says:  error: MongooseServerSelectionError: connect ECONNREFUSED 172.18.0.9:57398
Strangely, after 20 seconds (when the db is actually up and running) it says:  error: MongooseServerSelectionError: connect ECONNREFUSED 127.0.0.1:57398

Don't understand why IP address is changing to 127.0.0.1 after db is up and running.

**Reply by hasezoey** [reply]
> Don't understand why IP address is changing to 127.0.0.1 after db is up and running.

i also have no clue, maybe it is just a display error, because i have also observed this behavior in other issues, but still have not found a reason for it

> service 1 (which creates db replicaset): IP address: 172.18.0.9
> service 2(which consumes the db): IP address: 172.18.0.7

if i remember correctly, you dont need to deal with ip addresses, just use the container names (works like DNS)

in this case if your container name for the db container is `service-db`, then in the second instead of the IP, but in `service-db` (like `mongodb://service-db:57398`)

> To give some more info, I am facing this issue when I am using docker-compose

if your "problems" is just about *startup* then you should be able to use docker-compose config option [`depends_on`](https://docs.docker.com/compose/compose-file/compose-file-v3/#depends_on) which basically only starts the container that this field is on when all containers in the array are in state "running"

---

if it still does not work, maybe check if the wanted port is actually exposed in the docker-compose config (like `ports: - 57398:57398`)
